python/pyspark/ml/functions.py

```diff
@@ -106,6 +117,556 @@ def array_to_vector(col: Column) -> Column:
+            # x1.shape = [batch_size, 4]
+            # x2.shape = [batch_size, 3]
+            # outputs.shape = [batch_size], result_type = FloatType()
+            return np.sum(x1, axis=1) + np.sum(x2, axis=1)
+
+        return predict
+
+    sum_cols = predict_batch_udf(
+        multi_sum_fn,
+        return_type=FloatType(),
+        batch_size=5,
+        input_tensor_shapes=[[4], [3]],
+    )
+
+    df.withColumn("sum", sum_cols("t1", "t2")).show(5)
+    # +--------------------+------------------+-----+
+    # |                  t1|                t2|  sum|
+    # +--------------------+------------------+-----+
+    # |[0.0, 1.0, 2.0, 3.0]|   [0.0, 1.0, 2.0]|  9.0|
+    # |[4.0, 5.0, 6.0, 7.0]|   [4.0, 5.0, 6.0]| 37.0|
+    # |[8.0, 9.0, 10.0, ...|  [8.0, 9.0, 10.0]| 65.0|
+    # |[12.0, 13.0, 14.0...|[12.0, 13.0, 14.0]| 93.0|
+    # |[16.0, 17.0, 18.0...|[16.0, 17.0, 18.0]|121.0|
+    # +--------------------+------------------+-----+
+
+    # Note that some models can provide multiple outputs.  These can be returned as a dictionary
+    # of named values, which can be represented in columnar (or row-based) formats.
+
+    def multi_sum_fn():
+        def predict_columnar(x1: np.ndarray, x2: np.ndarray) -> Mapping[str, np.ndarray]:
+            # x1.shape = [batch_size, 4]
+            # x2.shape = [batch_size, 3]
+            return {
+                "sum1": np.sum(x1, axis=1),
+                "sum2": np.sum(x2, axis=1)
+            }  # return_type = StructType()
+
+        return predict_columnar
+
+    sum_cols = predict_batch_udf(
+        multi_sum_fn,
+        return_type=StructType([
+            StructField("sum1", FloatType(), True),
+            StructField("sum2", FloatType(), True)
+        ])
+        batch_size=5,
+        input_tensor_shapes=[[4], [3]],
+    )
+
+    df.withColumn("preds", sum_cols("t1", "t2")).select("t1", "t2", "preds.*").show(5)
+    # +--------------------+------------------+----+----+
+    # |                  t1|                t2|sum1|sum2|
+    # +--------------------+------------------+----+----+
+    # |[0.0, 1.0, 2.0, 3.0]|   [0.0, 1.0, 2.0]| 6.0| 3.0|
+    # |[4.0, 5.0, 6.0, 7.0]|   [4.0, 5.0, 6.0]|22.0|15.0|
+    # |[8.0, 9.0, 10.0, ...|  [8.0, 9.0, 10.0]|38.0|27.0|
+    # |[12.0, 13.0, 14.0...|[12.0, 13.0, 14.0]|54.0|39.0|
+    # |[16.0, 17.0, 18.0...|[16.0, 17.0, 18.0]|70.0|51.0|
+    # +--------------------+------------------+----+----+
+    ```
+
+    .. versionadded:: 3.4.0
+
+    Parameters
+    ----------
+    predict_batch_fn : Callable[[],
+        Callable[..., np.ndarray | Mapping[str, np.ndarray] | List[Mapping[str, np.dtype]] ]
+        Function which is responsible for loading a model and returning a `predict` function which
+        takes one or more numpy arrays as input and returns either a numpy array (for a single
+        output), a dictionary of named numpy arrays (for multiple outputs), or a row-oriented list
+        of dictionaries (for multiple outputs).
```

Comment:
For the row-oriented list case, let's document that each element in dict must be a scaler value or one dimension array, otherwise not supported.